src/analysis/ml.py

```diff
@@ -152,27 +175,92 @@ def plot_dendrogram(model, **kwargs):
     """
     Performs hierarchical agglomerative clustering on the dataframe,
-    creates a dendrogram of the resulting tree,
+    creates a dendrogram of the resulting tree USING SEABORN and SCKIT-LEARN for the cluster groups,
     and saves the dendrogram and the cluster labels of said dendrogram in separate files.
     @param dataframe: binary dataframe of edge comparison between algorithms from summarize_networks
     @param output_png: the file name to save the dendrogram image
     @param output_file: the file name to save the clustering labels
     """
+
+    if linkage not in linkage_methods:
+        raise ValueError(f"linkage={linkage} must be one of {linkage_methods}")
+    if linkage == "ward":
+        if metric != "euclidean":
+            print("For linkage='ward', the metric must be 'euclidean'; setting metric = 'euclidean")
+            metric = "euclidean"
+    if metric not in distance_metrics:
+        raise ValueError(f"metric={metric} must be one of {distance_metrics}")
+        
     X = dataframe.reset_index(drop=True)
-    X = X.transpose()
-    model = AgglomerativeClustering(distance_threshold=0.5, n_clusters=None)
+    column_names = X.head()
+    column_names = [element.split('-')[1] for element in column_names]
+    X = X.transpose() 
+
+    # creating the colors per algorithms
+    custom_palette = sns.color_palette("tab10", len(column_names))
+    label_color_map = {label: color for label, color in zip(column_names, custom_palette)}
+    row_colors = pd.Series(column_names, index=X.index).map(label_color_map)
+
+    #plotting the seaborn figure
+    plt.figure(figsize=(10, 7))
+    clustergrid = sns.clustermap(X, metric=metric, method=linkage, row_colors=row_colors, col_cluster=False)
+    clustergrid.ax_heatmap.remove()
+    clustergrid.cax.remove()
+    clustergrid.ax_row_dendrogram.set_visible(True)
+    clustergrid.ax_col_dendrogram.set_visible(False)
+    legend_labels = [plt.Rectangle((0, 0), 0, 0, color=label_color_map[label]) for label in label_color_map]
+    plt.legend(legend_labels, label_color_map.keys(), bbox_to_anchor=(1.02, 1), loc='upper left')
+    
+    make_required_dirs(output_png)
+    plt.savefig(output_png, bbox_inches="tight")
+
+    # getting the label of which group each algorithm combination falls under
+    model = AgglomerativeClustering(linkage=linkage, affinity=metric,distance_threshold=0.5, n_clusters=None)
     model = model.fit(X)
+    columns = dataframe.columns.tolist()
+    data = {'algorithm': columns, 'labels': model.labels_}
+    df = pd.DataFrame(data)
+    make_required_dirs(output_file)
+    df.to_csv(output_file, sep='\t', index=False)
+
 
+def hac_horizontal(dataframe: pd.DataFrame, output_png: str, output_file: str, linkage='ward', metric='euclidean'):
+    """
+    Performs hierarchical agglomerative clustering on the dataframe,
+    creates a dendrogram of the resulting tree USING SCKIT LEARN and makes cluster groups USING SCKIT LEARN,
```

Comment:
Let's avoid all caps. Isn't the dendrogram from scipy and the clustering from scikit-learn?